src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs

```diff
@@ -24,71 +34,184 @@ public static int CacheSize
+                        s_cacheFirst.Next = entry;
+                        entry.Previous = s_cacheFirst;
                     }
+                    s_cacheFirst = entry;
+
+                    s_cacheCount++;
+                    if (s_cacheCount >= CacheDictionarySwitchLimit)
+                    {
+                        if (s_cacheCount == CacheDictionarySwitchLimit)
+                            FillCacheDictionary();
+                        else
+                            s_cache.Add(key, entry);
+                        SysDebug.Assert(s_cacheCount == s_cache.Count);
+                    }
+
+                    // update last in linked list:
+                    if (s_cacheLast == null)
+                    {
+                        s_cacheLast = entry;
+                    }
+                    else if (s_cacheCount > s_cacheSize) // remove last
+                    {
+                        CachedCodeEntry last = s_cacheLast;
+                        if (s_cacheCount >= CacheDictionarySwitchLimit)
+                        {
+                            SysDebug.Assert(s_cache[last.Key] == s_cacheLast);
+                            s_cache.Remove(last.Key);
+                        }
+
+                        SysDebug.Assert(last.Previous == null);
+                        SysDebug.Assert(last.Next != null);
+                        SysDebug.Assert(last.Next.Previous == last);
+                        last.Next.Previous = null;
+                        s_cacheLast = last.Next;
+                        s_cacheCount--;
+                    }
+
                 }
+                return entry;
             }
+        }
 
-            return null;
+        private void FillCacheDictionary()
+        {
+            s_cache.Clear();
+            CachedCodeEntry next = s_cacheFirst;
+            while (next != null)
+            {
+                s_cache.Add(next.Key, next);
+                next = next.Previous;
+            }
         }
 
-        /// <summary>
-        /// Add current code to the cache
-        /// </summary>
-        private CachedCodeEntry CacheCode(CachedCodeEntryKey key)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // Unprofitable inline - JIT overly pessimistic
+        private static bool TryGetCacheValue(CachedCodeEntryKey key, out CachedCodeEntry entry)
         {
-            CachedCodeEntry newcached = null;
+            if (s_cacheCount >= CacheDictionarySwitchLimit)
+            {
+                SysDebug.Assert((s_cacheFirst != null && s_cacheLast != null && s_cache.Count > 0) || 
+                                (s_cacheFirst == null && s_cacheLast == null && s_cache.Count == 0), 
+                                "Linked list and Dict should be synchronized");
+                return s_cache.TryGetValue(key, out entry);
+            }
 
-            lock (s_livecode)
+            return TryGetCacheValueSmall(key, out entry);
```

Comment:
Attributing methods with AggressiveInlining can be very dangerous and regress perf in corner cases which are usually not tested. I factored out the retrieval for small caches so that the the caller's native size stays sane.

Reply:
IL Size is now 30 bytes.